File: compress/src/java/compresslib/Rle.java
package compresslib;

/**
 * 런 길이 부호 — SPEC §3.
 *
 * <p>정한 값 셋이 출력 바이트를 바꾼다: 문턱 3, 런 상한 128, 리터럴
 * 상한 128. 리터럴 묶음을 min(j+r, i+128) 로 자르는 것이 특히 중요하다
 * — 안 자르면 129바이트 묶음이 나오는데, 제어 바이트에 안 들어가는
 * 길이라 못 푼다.
 */
public final class Rle {
  public static final int RUN_MIN = 3;
  public static final int RUN_MAX = 128;
  public static final int LIT_MAX = 128;
  public static final int RESERVED = 128;

  private Rle() {}

  static int runAt(byte[] src, int i, int limit) {
    byte b = src[i];
    int j = i + 1;
    int end = Math.min(src.length, i + limit);
    while (j < end && src[j] == b) {
      j++;
    }
    return j - i;
  }

  public static void pack(byte[] src, ByteBuf out) {
    int i = 0;
    int n = src.length;
    while (i < n) {
      int run = runAt(src, i, RUN_MAX);
      if (run >= RUN_MIN) {
        out.push(257 - run);
        out.push(src[i] & 0xFF);
        i += run;
        continue;
      }
      int j = i;
      while (j < n && j - i < LIT_MAX) {
        int r = runAt(src, j, RUN_MAX);
        if (r >= RUN_MIN) {
          break;
        }
        j = Math.min(j + r, i + LIT_MAX);
      }
      out.push(j - i - 1);
      out.extend(src, i, j);
      i = j;
    }
  }

  /** 푼 바이트와 다음 위치. 위치는 pos[0] 으로 돌려준다. */
  public static byte[] unpack(byte[] src, int[] pos, int want) {
    ByteBuf out = new ByteBuf();
    int n = src.length;
    int p = pos[0];
    while (out.size() < want) {
      if (p >= n) {
        throw new CodecException("PackBits 가 잘렸다");
      }
      int c = src[p++] & 0xFF;
      if (c == RESERVED) {
        throw new CodecException("제어 128 은 쓰지 않는다");
      }
      if (c < RESERVED) {
        int k = c + 1;
        if (p + k > n) {
          throw new CodecException("리터럴 묶음이 잘렸다");
        }
        out.extend(src, p, p + k);
        p += k;
      } else {
        int k = 257 - c;
        if (p >= n) {
          throw new CodecException("런 묶음이 잘렸다");
        }
        for (int j = 0; j < k; j++) {
          out.push(src[p] & 0xFF);
        }
        p++;
      }
    }
    if (out.size() != want) {
      throw new CodecException("푼 길이가 헤더와 다르다");
    }
    pos[0] = p;
    return out.bytes();
  }

  public static byte[] encode(byte[] src) {
    ByteBuf out = new ByteBuf();
    Varint.put(out, src.length);
    pack(src, out);
    return out.bytes();
  }

  public static byte[] decode(byte[] src) {
    int[] h = Varint.getLength(src, 0);
    int[] pos = {h[1]};
    byte[] out = unpack(src, pos, h[0]);
    if (pos[0] != src.length) {
      throw new CodecException("뒤에 남은 바이트가 있다");
    }
    return out;
  }

  // 0런 부호 (SPEC §3.2) — bzip2 의 RUNA/RUNB. 13번 모듈이 쓴다.
  public static final int RUN_A = 0;
  public static final int RUN_B = 1;

  public static int[] zeroRunEncode(int[] syms) {
    java.util.ArrayList<Integer> out = new java.util.ArrayList<>();
    int i = 0;
    int n = syms.length;
    while (i < n) {
      if (syms[i] != 0) {
        out.add(syms[i] + 1);
        i++;
        continue;
      }
      int j = i;
      while (j < n && syms[j] == 0) {
        j++;
      }
      long length = (long) (j - i) + 1;
      while (length > 1) {
        out.add((length & 1) == 1 ? RUN_B : RUN_A);
        length >>>= 1;
      }
      i = j;
    }
    return out.stream().mapToInt(Integer::intValue).toArray();
  }

  public static int[] zeroRunDecode(int[] syms) {
    java.util.ArrayList<Integer> out = new java.util.ArrayList<>();
    int i = 0;
    int n = syms.length;
    while (i < n) {
      if (syms[i] > 1) {
        out.add(syms[i] - 1);
        i++;
        continue;
      }
      long run = 0;
      long weight = 1;
      while (i < n && syms[i] <= 1) {
        run += (long) (syms[i] + 1) * weight;
        weight <<= 1;
        i++;
      }
      for (long j = 0; j < run; j++) {
        out.add(0);
      }
    }
    return out.stream().mapToInt(Integer::intValue).toArray();
  }
}

File: compress/src/py/compresslib/rle.py
# -*- coding: utf-8 -*-
"""런 길이 부호 — SPEC §3.

같은 바이트가 이어지면 "그 바이트와 개수" 로 적는다. 가장 단순한
압축이고, 그래서 한계도 가장 먼저 보인다 — 런이 없으면 반드시 커진다.

여기 둘이 있다.
  PackBits  바이트 단위. 리터럴 묶음과 런 묶음을 제어 바이트로 가른다
  0런 부호  MTF 뒤의 0 무더기를 위한 것. bzip2 가 쓰는 RUNA/RUNB 다

PackBits 에서 정한 것 셋: 문턱 3, 런 상한 128, 리터럴 상한 128.
셋 다 출력 바이트를 바꾸는 결정이라 SPEC §3.1 에 못 박혀 있다.
시간·공간 모두 O(입력 길이).
"""
from compresslib import varint

RUN_MIN = 3
RUN_MAX = 128
LIT_MAX = 128
# 제어 128 은 안 쓴다. 애플 원본은 "아무 일도 안 함" 으로 넘어가지만,
# 같은 입력을 두 가지 바이트로 적을 수 있는 부호기는 골든 벡터를 못
# 가진다.
RESERVED = 128


def _run_at(src, i, cap=RUN_MAX):
    """i 에서 시작하는 같은 바이트의 길이 (cap 까지)."""
    b = src[i]
    j = i + 1
    end = min(len(src), i + cap)
    while j < end and src[j] == b:
        j += 1
    return j - i


def pack(src):
    """PackBits 묶음들. 헤더는 붙이지 않는다."""
    out = bytearray()
    i, n = 0, len(src)
    while i < n:
        run = _run_at(src, i)
        if run >= RUN_MIN:
            out.append(257 - run)
            out.append(src[i])
            i += run
            continue
        # 리터럴 — 런이 시작되는 자리까지, 또는 상한까지
        j = i
        while j < n and (j - i) < LIT_MAX:
            r = _run_at(src, j)
            if r >= RUN_MIN:
                break
            # 상한을 넘겨 잡으면 129바이트 묶음이 나온다. 제어 바이트에
            # 안 들어가는 길이라, 아무도 못 푸는 파일이 된다.
            j = min(j + r, i + LIT_MAX)
        out.append(j - i - 1)
        out += src[i:j]
        i = j
    return bytes(out)


def unpack(src, pos, want):
    """묶음을 풀어 want 바이트를 만든다. (바이트, 다음 위치)."""
    out = bytearray()
    n = len(src)
    while len(out) < want:
        if pos >= n:
            raise ValueError('PackBits 가 잘렸다')
        c = src[pos]
        pos += 1
        if c == RESERVED:
            raise ValueError('제어 128 은 쓰지 않는다')
        if c < RESERVED:
            k = c + 1
            if pos + k > n:
                raise ValueError('리터럴 묶음이 잘렸다')
            out += src[pos:pos + k]
            pos += k
        else:
            k = 257 - c
            if pos >= n:
                raise ValueError('런 묶음이 잘렸다')
            out += bytes([src[pos]]) * k
            pos += 1
    if len(out) != want:
        raise ValueError('푼 길이가 헤더와 다르다: %d != %d'
                         % (len(out), want))
    return bytes(out), pos


def encode(src):
    return varint.put(len(src)) + pack(src)


def decode(src):
    n, pos = varint.get_length(src)
    out, pos = unpack(src, pos, n)
    if pos != len(src):
        raise ValueError('뒤에 남은 바이트가 있다')
    return out


# ---------------------------------------------------------------- 0런
# 부호 SPEC §3.2. MTF 를 거친 열은 0 이 압도적으로 많다. 그 0 무더기를
# 이진법의 **자릿수** 로 적는다 — RUNA 가 1, RUNB 가 2 를 나타내는 쌍대
# 이진법이라 앞자리 0 이 없고, 그래서 같은 길이를 두 가지로 적을 수
# 없다.
RUNA = 0
RUNB = 1


def zero_run_encode(symbols):
    """MTF 기호 열 → 0런이 접힌 열. 0 아닌 기호는 한 칸씩 밀린다."""
    out = []
    i, n = 0, len(symbols)
    while i < n:
        if symbols[i] != 0:
            out.append(symbols[i] + 1)
            i += 1
            continue
        j = i
        while j < n and symbols[j] == 0:
            j += 1
        length = (j - i) + 1
        while length > 1:
            out.append(RUNB if (length & 1) else RUNA)
            length >>= 1
        i = j
    return out


def zero_run_decode(symbols):
    out = []
    i, n = 0, len(symbols)
    while i < n:
        if symbols[i] > 1:
            out.append(symbols[i] - 1)
            i += 1
            continue
        run, weight = 0, 1
        while i < n and symbols[i] <= 1:
            run += (symbols[i] + 1) * weight
            weight <<= 1
            i += 1
        out += [0] * run
    return out
